transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java

```diff
@@ -35,6 +36,18 @@
     private final String name;
     private boolean removed;
 
+    // This will set to true once the ChannelHandler.handlerAdded(...) method was called. We need to keep track of this
+    // to ensure we will never call another ChannelHandler method before handlerAdded(...) was called to guard against
+    // ordering issues. handlerAdded(...) MUST be the first method that is called per handler when become part of the
+    // ChannelPipeline in all cases. Not doing so may lead to unexpected side-effects as ChannelHandler implementations
+    // may need to do initialization steps before a ChannelHandler can be used.
+    //
+    // See:
+    //   - https://github.com/netty/netty/issues/4705
+    //
+    // No need to mark volatile as this will be made visible as next/prev is volatile.
+    private boolean handlerAdded;
```

Comment:
Does this need to be a bit more generic to represent the implicit ordering dependencies between other methods? Are there other methods that could be called from other threads which may violate the order for other methods? For example does this need to be a `byte state` (or something similar) which each method needs to check if they can execute immediately or dispatch a task to the event loop to run later?